### PR TITLE
Linux compilation support

### PR DESCRIPTION
**Pull request by Akatsuki57**
Hi, this time, it's don't break your Windows tools-chain.
Thx @pbomta for cmake tips.

This pull request allows to compile d2tm on linux ubuntu 21.10 without any other modification.
Launch 
sh install_dependancies_ubuntu.sh
mkdir build 
cd build
cmake ..
make -j4

### CODE REVIEW

File: player/brains/superweapon/cPlayerBrainFremenSuperWeapon.cpp [hunk 1]
```diff
@@ -32,10 +32,12 @@ namespace brains {
...
         }
 
-        char msg[255];
-        sprintf(msg, "cPlayerBrainFremenSuperWeapon::think() AI[%d] - found %d idle unit(s) to attack with.",
-                player->getId(), ids.size());
-        logbook(msg);
+        // char msg[255];
+        // sprintf(msg, "cPlayerBrainFremenSuperWeapon::think() AI[%d] - found %d idle unit(s) to attack with.",
+        //         player->getId(), ids.size());               
+        // logbook(msg);
```
Comment (pbomta) [inline player/brains/superweapon/cPlayerBrainFremenSuperWeapon.cpp:38]: No need to leave comments. Just delete code. :) 
We can look in the git history if we need to.

File: dependencies/alfont/CMakeLists.txt [hunk 2]
```diff
@@ -31,6 +31,10 @@ set(FREETYPE_SOURCES
     "freetype/src/autofit/autofit.c"
 )
 
+if (UNIX)
+  add_definitions(-DALFONT_LINUX)
```
Comment (pbomta) [inline dependencies/alfont/CMakeLists.txt:35]: Modern CMake prefers target properties. How about my proposal: https://github.com/stefanhendriks/Dune-II---The-Maker/pull/445/files ?

**Review by pbomta — APPROVED** [review]

**Comment by Akatsuki57** [comment]
Whoops ! I worked yesterday on removing sprintf in software. I sped up the use of fmt::format
I had wanted to create a new pull request but my work was integrated on this flow.
It bothers me a bit because that wasn't the end goal of this pull request.

File: CMakeLists.txt [hunk 3]
```diff
@@ -2,7 +2,27 @@ cmake_minimum_required(VERSION 3.21)
...
 set(CMAKE_CXX_STANDARD 20)
 
+message("---------------------------------------------")
+message(" Project d2tm")
+message("---------------------------------------------")
+
 project(d2tm)
+SET(D2TM_MAJOR "0")
+SET(D2TM_MINOR "6")
+SET(D2TM_DAYLY "X")
```
Comment (stefanhendriks) [inline CMakeLists.txt:12]: ooh can we now also use the `git sha`?

**Comment by stefanhendriks** [comment]
`master` has been updated , please make sure to resolve conflicts and check if it works. If 👍 then we can merge 🎉 

Would be nice to have a GH action for linux btw, so we can have a CI step to tell us things keep working there.

**Comment by stefanhendriks** [comment]
Also plz check @pbomta's comments

File: cGame_logic.cpp [hunk 4]
```diff
@@ -55,7 +56,7 @@ cGame::cGame() : m_timeManager(*this) {
     m_iniScreenWidth = -1;
     m_iniScreenHeight = -1;
 
-    m_version = "0.6.x";
+    m_version = D2TM_VERSION;
```
Comment (pbomta) [inline cGame_logic.cpp:59]: Why move this to CMake?
Many libraries have their version in the source code. Usually a version.h or something.

File: cGame_logic.cpp [hunk 5]
```diff
@@ -688,16 +690,16 @@ bool cGame::setupGame() {
 
     game.init(); // Must be first! (loads game.ini file at the end, which is required before going on...)
 
-    logger->logHeader("Dune II - The Maker");
+    logger->logHeader(D2TM_APP_NAME);
```
Comment (pbomta) [inline cGame_logic.cpp:693]: Let's leave the string here, that's easier than using a define.
Also, these defines are available in each and every source file, which goes against the principle of data hinding.

File: config.h [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+// do not edit this file
+
+
+#ifndef _CONFIG_H_
+#define _CONFIG_H_
+
+#define D2TM_PACKAGE  "d2tm"
+#define D2TM_VERSION  "0.6.X"
+#define D2TM_APP_NAME "Dune II - The Maker"
+#define D2TM_OS       "Linux OS"
```
Comment (pbomta) [inline config.h:11]: defines do not respect namespaces. OK, we don't use namespaces yet in this project, but modern C++ uses constexpr char* for strings like this.

File: config.h.in [hunk 7]
```diff
@@ -0,0 +1,13 @@
+// the configured options and settings 
+// do not edit this file
```
Comment (pbomta) [inline config.h.in:2]: Oh, no... Sorry to rain on your parade, but I used libraries that generated source files using the buildsystem. It was always a mess. Also, I don't see the reason for this. It only serves to make things complicated.

File: gamestates/cSetupSkirmishGameState.cpp [hunk 8]
```diff
@@ -640,10 +635,7 @@ void cSetupSkirmishGameState::prepareSkirmishGameToPlayAndTransitionToCombatStat
         int maxDistance = worms * 32; // 128 / 4
         int wormCell = map.getRandomCell();
         int failures = 0;
-        char msg[255];
-        sprintf(msg, "Skirmish game with %d sandworms, minDistance %d, maxDistance %d", worms, minDistance,
-                maxDistance);
-        logbook(msg);
+        logbook(fmt::format("Skirmish game with {} sandworms, minDistance {}, maxDistance {}", worms, minDistance, maxDistance));
```
Comment (pbomta) [inline gamestates/cSetupSkirmishGameState.cpp:638]: All the fmt conversion is appreciated though! I wish it was a separate commit. Then we could have merged it already.

File: map/cMapEditor.cpp [hunk 9]
```diff
@@ -1,5 +1,8 @@
+#include "../include/d2tmh.h"
```
Comment (pbomta) [inline map/cMapEditor.cpp:1]: Nope. We're trying to get rid of this header.

Also, always include the corresponding header first. Then the compilation run checks if that header can be included without caring about include order.

**Review by pbomta — CHANGES_REQUESTED** [review]
Let's not generate code from the buildsystem.

File: map/cMapEditor.cpp (hunk 9 above)
Comment (Akatsuki57) [reply map/cMapEditor.cpp:1]: Not me this include.
It will be here because code change.

**Comment by Akatsuki57** [comment]
Cancel last comit i have make and I will continue to work.  There is so many think to do. I want to build another pull request but I miss something and git get all in one.

**Comment by stefanhendriks** [comment]
Having read feedback from @pbomta I agree with his points. Lets not do the versioning like that, but keep the rest. After that, lets merge this first before creating more changes (in this PR), so we can create a new PR for the new changes.

File: config.h.in (hunk 7 above)
Comment (Akatsuki57) [reply config.h.in:2]: I'm not stupid

File: cGame_logic.cpp (hunk 4 above)
Comment (Akatsuki57) [reply cGame_logic.cpp:59]: I don't know. Maybe it's easier than existing ?

File: config.h.in (hunk 7 above)
Comment (pbomta) [reply config.h.in:2]: What in my comment makes you say that?

**Comment by Akatsuki57** [comment]
PR done !

**Comment by stefanhendriks** [comment]
As discussed on Discord, I have taken this PR and split it into two:

1. For linux compilation (including GH action for Ubuntu)
2. Fmt fixes (https://github.com/stefanhendriks/Dune-II---The-Maker/pull/454 in progress)

Once the 2nd PR is merged, we can close this PR. The changes have been merged as said.

**Pull request closed without merging**